consumer-test/TestRunnerThread.groovy

```diff
@@ -0,0 +1,137 @@
+
+        //Commit the changed version because artifactSetVersion expects the git repo not to be dirty
+        executeShell(["git", "-C", "${testCaseWorkspace}", "commit", "--all",
+                        "--author=piper-testing-bot <piper-testing-bot@example.com>",
+                        "--message=Set piper lib version for test"])
+
+        executeShell("docker run -v /var/run/docker.sock:/var/run/docker.sock " +
+            "-v ${System.getenv('PWD')}/${testCaseWorkspace}:/workspace -v /tmp " +
+            "-e CASC_JENKINS_CONFIG=/workspace/jenkins.yml -e CX_INFRA_IT_CF_USERNAME " +
+            "-e CX_INFRA_IT_CF_PASSWORD -e BRANCH_NAME=${testCase} ppiper/jenkinsfile-runner")
+
+        println "*****[INFO] Test case '${uniqueName}' finished successfully.*****"
+        printOutput()
+    }
+
+    // Configure path to library-repository under test in Jenkins config
+    private void addJenkinsYmlToWorkspace() {
+        def sourceFile = 'jenkins.yml'
+        def sourceText = new File(sourceFile).text.replaceAll(
+            '__REPO_SLUG__', ConsumerTestUtils.repositoryUnderTest)
+        def target = new File("${testCaseWorkspace}/${sourceFile}")
+        target.write(sourceText)
+    }
+
+    // Force usage of library version under test by setting it in the Jenkinsfile,
+    // which is then the first definition and thus has the highest precedence.
+    private void manipulateJenkinsfile() {
+        def jenkinsfile = new File("${testCaseWorkspace}/Jenkinsfile")
+        def manipulatedText =
+            "@Library(\"piper-library-os@${ConsumerTestUtils.libraryVersionUnderTest}\") _\n" +
+                jenkinsfile.text
+        jenkinsfile.write(manipulatedText)
+    }
+
+    private void executeShell(command) {
+        def startOfCommandString = "Shell command: '${command}'\n"
+        stdOut << startOfCommandString
+        stdErr << startOfCommandString
+
+        currentProcess = command.execute()
+        currentProcess.waitForProcessOutput(stdOut, stdErr)
+
+        exitCode = currentProcess.exitValue()
+
+        def endOfCommandString = "*****Command execution finished with exit code ${exitCode}" +
+            ".*****\n\n"
+        stdOut << endOfCommandString
+        stdErr << endOfCommandString
+
+        currentProcess = null
+
+        if (this.exitCode > 0) {
+            synchronized (this) {
+                try {
+                    wait() // for other threads to print their log first
+                    // then it is interrupted
+                } catch (InterruptedException e) {
+                    printOutput()
+                    notifyGithub("failure", "Consumer test ${uniqueName} failed.")
+                    exitPrematurely(exitCode, "Consumer test ${uniqueName} failed, aborted!")
+                }
+            }
+        }
+    }
+
+    void printOutput() {
+        println "\n[INFO] Standard output from test case ${uniqueName}:"
+        stdOut?.eachLine { line, i ->
```

Comment:
Only if the print method is called very early in the game, but that should not happen.

Reply:
> is called very early in the game

No. The variables are assigned with a value directly at declaration. They can not be null (unless somebody sets it to `null` explictly, which I do not see (... having it final would help here). Even the variables would be initialized inside constructor they cannot be null (again: if not explicitly set to null).

This applies also elsewhere below, also for stderr. The point behind is: in case the variable is null I prefer to get the corresponding NPE, rather than having the log output silently swallowed ...